### PR TITLE
TF-1311:  Hide Team-Mailboxes in mailbox menu

### PR DESCRIPTION
**Pull request by nqhhdev**
- Set subscribeMailbox in UseCase and implement in Repository
 - Implement `SetMailboxMethod` in `MailboxApi` to update property `isSubscribed` = false to hide mailbox
 
 - Demo `WEB`
 
https://user-images.githubusercontent.com/99852347/217490928-57b91160-0bd4-414d-b887-e48d8567e557.mov

- Demo `Mobile`

[mobile.webm](https://user-images.githubusercontent.com/99852347/217492006-d934d5e6-d8e9-4747-9819-30df271ee2d0.webm)

### CODE REVIEW

File: lib/features/mailbox/domain/usecases/subscribe_mailbox_interactor.dart [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class SearchMailboxInteractor {
+  final MailboxRepository _mailboxRepository;
+
+  SearchMailboxInteractor(this._mailboxRepository);
+
+  Stream<Either<Failure, Success>> execute(AccountId accountId, SubscribeMailboxRequest request) async* {
+    try {
+      yield Right<Failure, Success>(LoadingSubscribeMailbox());
+
+      final resultList = await _mailboxRepository.subscribeMailbox(accountId, request);
```
Comment (dab246) [inline lib/features/mailbox/domain/usecases/subscribe_mailbox_interactor.dart:18]: rename to `result` or `isSuccess`

File: lib/features/mailbox/domain/model/mailbox_subscribe_state.dart [hunk 2]
```diff
@@ -0,0 +1,13 @@
+enum MailboxSubscribeState {
+  enabled,
+  disabled;
+  
+  String get keyValue {
```
Comment (dab246) [inline lib/features/mailbox/domain/model/mailbox_subscribe_state.dart:5]: If you don't use it, delete it.

**Comment by dab246** [comment]
Screenshot 2023-02-07 at 15 26 40

Screenshot 2023-02-07 at 15 26 48

- The left margin is a bit big

File: lib/features/mailbox/domain/state/subscribe_mailbox_state.dart [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import 'package:core/core.dart';
+import 'package:tmail_ui_user/features/base/state/ui_action_state.dart';
+import 'package:jmap_dart_client/jmap/core/state.dart' as jmap;
+
+
```
Comment (dab246) [inline lib/features/mailbox/domain/state/subscribe_mailbox_state.dart:5]: remove excess line

File: lib/features/mailbox/domain/state/subscribe_mailbox_state.dart [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+class SubscribeMailboxSuccess extends UIActionState {
+
+  SubscribeMailboxSuccess({
+    jmap.State? currentEmailState,
+    jmap.State? currentMailboxState,
+  }) : super(currentEmailState, currentMailboxState);
+
+  @override
+  List<Object?> get props => [];
```
Comment (dab246) [inline lib/features/mailbox/domain/state/subscribe_mailbox_state.dart:16]: should fill in all properties to compare instance

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 5]
```diff
@@ -1206,4 +1217,19 @@ class MailboxController extends BaseMailboxController {
...
+
+  void _refreshMailboxTreeHasSubscribed(List<PresentationMailbox> mailboxList) async {
+    final _mailboxList = mailboxList.where((mailbox) => mailbox.isSubscribed?.value == true).toList();
+    await refreshTree(_mailboxList);
+  }
+
+  void subscribeMailboxAction(PresentationMailbox mailboxSelected) {
+    final _accountId = mailboxDashBoardController.accountId.value;
+    if(_accountId != null) {
+      consumeState(_subscribeMailboxInteractor.execute(_accountId, SubscribeMailboxRequest(
```
Comment (dab246) [inline lib/features/mailbox/presentation/mailbox_controller.dart:1229]:
```suggestion
consumeState(_subscribeMailboxInteractor.execute(
   _accountId, 
   SubscribeMailboxRequest(
      mailboxSelected.id,
      MailboxSubscribeState.disabled,
    )
));
```

File: lib/features/mailbox/presentation/mailbox_view.dart [hunk 6]
```diff
@@ -331,11 +331,15 @@ class MailboxView extends GetWidget<MailboxController> {
 
   void _openBottomSheetSpamMenuAction(BuildContext context, PresentationMailbox mailbox) {
     final _spamActionsSupported = [
```
Comment (dab246) [inline lib/features/mailbox/presentation/mailbox_view.dart:333]: rename to `_mailboxActionsSupported `

File: lib/features/mailbox/presentation/mailbox_view.dart [hunk 7]
```diff
@@ -331,11 +331,15 @@ class MailboxView extends GetWidget<MailboxController> {
...
+      MailboxActions.disableMailbox
     ];
 
+    if (mailbox.isSpam) {
+      _spamActionsSupported.add(controller.mailboxDashBoardController.enableSpamReport
+        ? MailboxActions.disableSpamReport
+        : MailboxActions.enableSpamReport);
+    }
+
     final listContextSpamPopupMenuItemAction = _spamActionsSupported
```
Comment (dab246) [inline lib/features/mailbox/presentation/mailbox_view.dart:343]: rename to `listContextMailboxPopupMenuItemAction `

**Comment by dab246** [comment]
### Note
-  Team-Mailbox (parentId != null) and Folder-Mailbox => Option `Hide Mailbox` is `Activated`
- `DefaultMailbox` => Option `Hide Mailbox` is `Deactivated`
- Else not include

File: lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart [hunk 8]
```diff
@@ -15,7 +15,7 @@ typedef OnOpenMailboxFolderClick = void Function(MailboxNode);
 typedef OnSelectMailboxFolderClick = void Function(MailboxNode);
 typedef OnMenuActionClick = void Function(RelativeRect, MailboxNode);
 typedef OnDragItemAccepted = void Function(List<PresentationEmail>, PresentationMailbox);
-typedef OnLongPressSpamReport = void Function(MailboxNode);
+typedef OnLongPressMailboxFolderClick = void Function(MailboxNode);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart:18]: @@ what is this change?

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 9]
```diff
@@ -289,4 +291,32 @@ class MailboxAPI with HandleSetErrorMixin {
...
+        ..usings(setMailboxMethod.requiredCapabilities))
+      .build()
+      .execute();
+
+    final setMailboxResponse = response.parse<SetMailboxResponse>(
+        setMailboxInvocation.methodCallId,
+        SetMailboxResponse.deserialize);
+
+    return Future.sync(() async {
+      return setMailboxResponse?.updated?.isNotEmpty == true;
```
Comment (hoangdat) [inline lib/features/mailbox/data/network/mailbox_api.dart:317]: how about to compare it with `request.mailboxId.id`

File: lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart (hunk 8 above)
Comment (nqhhdev) [reply lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart:18]: This func has a purpose when users of LongPress on mobile will happen 2 cases
1. If it is `SpamMailbox` will show enableSpamReport/disableMailbox
2. If it is Team-Mailbox (`parentId`!= null) and Folder-Mailbox will show `hideMailbox`

So I changed the name for clarity

File: lib/features/base/base_mailbox_controller.dart [hunk 10]
```diff
@@ -149,7 +149,8 @@ abstract class BaseMailboxController extends BaseController {
       personalMailboxTree.value.root.childrenItems?.isNotEmpty ?? false;
   
   bool get teamMailboxesHasChild =>
-      teamMailboxesTree.value.root.childrenItems?.isNotEmpty ?? false;
+      (teamMailboxesTree.value.root.childrenItems?.isNotEmpty ?? false )
```
Comment (dab246) [inline lib/features/base/base_mailbox_controller.dart:152]: should be explicitly written as a function.
```
bool methodName() {
 return;
}
```

File: lib/features/mailbox/domain/model/subscribe_mailbox_request.dart [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import 'package:equatable/equatable.dart';
+import 'package:model/mailbox/presentation_mailbox.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subscribe_action_state.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subscribe_state.dart';
+
+class SubscribeMailboxRequest with EquatableMixin {
+
+  final PresentationMailbox mailbox;
```
Comment (dab246) [inline lib/features/mailbox/domain/model/subscribe_mailbox_request.dart:8]: IMO, web only need `mailboxId` to handle update `isSubscribed`

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 12]
```diff
@@ -1206,4 +1221,47 @@ class MailboxController extends BaseMailboxController {
...
+
+  void _refreshMailboxTreeHasSubscribed(List<PresentationMailbox> mailboxList) async {
+    final _mailboxList = mailboxList.where((mailbox) => mailbox.isSubscribed?.value == true).toList();
+    await refreshTree(_mailboxList);
+  }
+
+  void subscribeMailboxAction(SubscribeMailboxRequest subscribeMailboxRequest) {
+    final _accountId = mailboxDashBoardController.accountId.value;
+    if(_accountId != null) {
+      consumeState(_subscribeMailboxInteractor.execute(
```
Comment (dab246) [inline lib/features/mailbox/presentation/mailbox_controller.dart:1233]:
```suggestion
consumeState(_subscribeMailboxInteractor.execute(
    _accountId, 
    subscribeMailboxRequest
));
```

File: lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart [hunk 13]
```diff
@@ -192,6 +199,18 @@ class MailBoxFolderTileBuilder {
     }
   }
 
+  // double _buildSizedBoxWhenHasNotChildren() {
```
Comment (dab246) [inline lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart:202]: remove it

File: lib/features/base/base_mailbox_controller.dart [hunk 14]
```diff
@@ -148,8 +148,10 @@ abstract class BaseMailboxController extends BaseController {
   bool get personalMailboxHasChild =>
       personalMailboxTree.value.root.childrenItems?.isNotEmpty ?? false;
   
-  bool get teamMailboxesHasChild =>
-      teamMailboxesTree.value.root.childrenItems?.isNotEmpty ?? false;
+  bool get teamMailboxesHasChild {
+    return (teamMailboxesTree.value.root.childrenItems?.isNotEmpty ?? false )
```
Comment (dab246) [inline lib/features/base/base_mailbox_controller.dart:152]:
```suggestion
return (teamMailboxesTree.value.root.childrenItems?.isNotEmpty ?? false)
```

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 15]
```diff
@@ -61,7 +63,7 @@ class MailboxAPI with HandleSetErrorMixin {
     return MailboxResponse(mailboxes: resultCreated?.list, state: resultCreated?.state);
   }
 
-  Set<CapabilityIdentifier> capabilitiesSupportedMailboxes(Session session, AccountId accountId) {
+  Set<CapabilityIdentifier> capabilitiesForGetMailboxMethod(Session session, AccountId accountId) {
```
Comment (hoangdat) [inline lib/features/mailbox/data/network/mailbox_api.dart:66]: why not?
```suggestion
Set<CapabilityIdentifier> _capabilitiesForGetMailboxMethod(Session session, AccountId accountId) {
```

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 16]
```diff
@@ -99,8 +101,10 @@ class MailboxAPI with HandleSetErrorMixin {
     final getMailboxUpdatedInvocation = jmapRequestBuilder.invocation(getMailboxUpdated);
     final getMailboxCreatedInvocation = jmapRequestBuilder.invocation(getMailboxCreated);
 
+    final capabilities = capabilitiesForGetMailboxMethod(session, accountId);
```
Comment (hoangdat) [inline lib/features/mailbox/data/network/mailbox_api.dart:104]: `capabilitiesForGetMailboxMethod` not make sense in this case

File: lib/features/mailbox/data/network/mailbox_api.dart (hunk 9 above)
Comment (hoangdat) [reply lib/features/mailbox/data/network/mailbox_api.dart:319]: why you resolve my question? :)))

File: lib/features/mailbox/domain/model/subscribe_mailbox_request.dart [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import 'package:equatable/equatable.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subscribe_action_state.dart';
+import 'package:tmail_ui_user/features/mailbox/domain/model/mailbox_subscribe_state.dart';
+
+class SubscribeMailboxRequest with EquatableMixin {
+
+  final MailboxId mailboxId;
+  final MailboxSubscribeState newState;
+  final MailboxSubscribeStateAction mailboxSubscribeStateAction;
```
Comment (hoangdat) [inline lib/features/mailbox/domain/model/subscribe_mailbox_request.dart:10]: what is the different?

**Comment by hoangdat** [comment]
image

Why the toast in android device is not clearly?

File: lib/features/mailbox/domain/model/subscribe_mailbox_request.dart (hunk 17 above)
Comment (nqhhdev) [reply lib/features/mailbox/domain/model/subscribe_mailbox_request.dart:10]: MailboxSubscribeState :
- Disable/enable
MailboxSubscribeStateAction :
- Subscribing/Undo

Comment (hoangdat) [reply lib/features/mailbox/domain/model/subscribe_mailbox_request.dart:10]: why we need all of these?

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 18]
```diff
@@ -1206,4 +1222,49 @@ class MailboxController extends BaseMailboxController {
     final _mailboxList = mailboxList.where((mailbox) => mailbox.isSubscribed?.value == true).toList();
     await buildTree(_mailboxList);
   }
+
+  void _refreshMailboxTreeHasSubscribed(List<PresentationMailbox> mailboxList) async {
+    final _mailboxList = mailboxList.where((mailbox) => mailbox.isSubscribed?.value == true).toList();
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:1227]: why not do it in the `interactor`?

**Review by dab246 — APPROVED** [review]
lgtm

File: lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart [hunk 19]
```diff
@@ -170,6 +176,7 @@ class MailBoxFolderTileBuilder {
           child: Material(
             color: Colors.transparent,
             child: InkWell(
+              onLongPress: () => print('123'),
```
Comment (dab246) [inline lib/features/mailbox/presentation/widgets/mailbox_folder_tile_builder.dart:179]: remove it

**Review by dab246 — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**